### PR TITLE
csi: add hooks in both Node and Controller Services

### PR DESCRIPTION
**Pull request by bertinatto**
CC @openshift/storage

### CODE REVIEW

**Comment by bertinatto** [comment]
/hold

This is a WIP design.

**Comment by jsafrane** [comment]
Should there be a similar hook for DaemonSet?

**Comment by bertinatto** [comment]
> Should there be a similar hook for DaemonSet?

Added, but no operator is using it ATM

**Comment by bertinatto** [comment]
/hold cancel

I've tested this with openshift/aws-ebs-csi-driver-operator#104 and it works.

**Comment by bertinatto** [comment]
@deads2k could you take a look, please?

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go [hunk 1]
```diff
@@ -89,14 +92,16 @@ func NewCSIDriverControllerServiceController(
 	deployInformer appsinformersv1.DeploymentInformer,
 	optionalConfigInformer configinformers.SharedInformerFactory,
 	recorder events.Recorder,
+	optionalDeploymentHooks ...func(*appsv1.Deployment) error,
```
Comment (jsafrane) [inline pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:95]: Would it make sense to either give owner of the controller a way how to poke the controller for resync (i.e. the owner has informer on secrets / config maps / whatnot and is able to enqueue "fake" event into CSIDriverControllerServiceController' queue) or pass informers to watch into the `NewCSIDriverControllerServiceController`?

I can see there is `ResyncEvery(time.Minute)` below, so this PR will work even without it.

Comment (bertinatto) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:95]: Would it make sense to address this in another PR?

Also, could you clarify a use case for poking the controller for resync?

Comment (jsafrane) [reply pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller.go:95]: Ack

File: pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller_test.go [hunk 2]
```diff
@@ -422,33 +469,6 @@ func TestSync(t *testing.T) {
 					withFalseConditions(conditionProgressing)),
 			},
 		},
-		{
-			// Deployment is fully deployed and its status is synced to CR
```
Comment (deads2k) [inline pkg/operator/csi/csidrivercontrollerservicecontroller/csi_driver_controller_service_controller_test.go:426]: this rubs me the wrong way.  Why did it need to be removed?

File: pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go [hunk 3]
```diff
@@ -63,6 +64,8 @@ type CSIDriverNodeServiceController struct {
 	operatorClient v1helpers.OperatorClient
 	kubeClient     kubernetes.Interface
 	dsInformer     appsinformersv1.DaemonSetInformer
+	// Optional hook functions to modify the DaemonSet.
+	optionalDaemonSetHooks []func(*appsv1.DaemonSet) error
```
Comment (deads2k) [inline pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go:68]: typed functions please. Also, make clear what happens on errors and where responsibility for identifying *which* function is failing lies in the error.

File: pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go [hunk 4]
```diff
@@ -116,6 +121,13 @@ func (c *CSIDriverNodeServiceController) sync(ctx context.Context, syncContext f
 	manifest := replacePlaceholders(c.manifest, opSpec)
 	required := resourceread.ReadDaemonSetV1OrDie(manifest)
 
+	for _, fn := range c.optionalDaemonSetHooks {
+		err := fn(required)
+		if err != nil {
+			return fmt.Errorf("error running hook function: %w", err)
```
Comment (deads2k) [inline pkg/operator/csi/csidrivernodeservicecontroller/csi_driver_node_service_controller.go:127]: you can at least give me an ordinal (`i`)

**Comment by deads2k** [comment]
I'm curious what you're trying to replace/mutate.  Some sorts of mutations are easier while you still have bytes (this is why we have the `replacePlaceholders`).

**Review by bertinatto — COMMENTED** [review]
> I'm curious what you're trying to replace/mutate. Some sorts of mutations are easier while you still have bytes (this is why we have the replacePlaceholders).

As more CSI operators use these control loops, we want to give them freedom to mutate the deployment/daemonset without adding placeholders to library-go, specially if the mutation is specific to that operator, like this one: https://github.com/openshift/aws-ebs-csi-driver-operator/pull/102

Most operators won't need this, and the default mutation approach should be the string replacements, but I figured for some situations mutating the object is better.

Does that make sense? Do you think another controller would be a better approach?

**Comment by deads2k** [comment]
> Does that make sense?

I understand the desire.  I'd be somewhat interested in what you're combining because I've regretted it in the past.  But I'm ok adding the option

**Pull request merged**